### PR TITLE
Rewrite inplace pass and fix gc bug

### PR DESCRIPTION
**Pull request by sneaxiy**

### CODE REVIEW

File: paddle/fluid/framework/details/inplace_op_pass.cc [hunk 1]
```diff
@@ -83,490 +78,373 @@ const std::string kInplacedOpWhiteList[] = { // NOLINT
...
+  PADDLE_ENFORCE(graph->Has(details::kMemOptSkipVars),
+                 "Graph should have attr %s", details::kMemOptSkipVars);
+  auto &mem_opt_whitelist = graph->Get<MemOptSkipVars>(kMemOptSkipVars);
+  for (const auto &var : mem_opt_whitelist) {
+    skip_vars_.emplace(var);
+  }
+
+  // 2. track the nodes which used by parameter server.
+  // these node can not be inplaced, otherwise trainer
+  // pserver can not find each other name.
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/inplace_op_pass.cc]: Tiny Grammar issue: "each other name" should be "each other's name"

File: paddle/fluid/framework/details/inplace_op_pass.cc [hunk 2]
```diff
@@ -83,490 +78,373 @@ const std::string kInplacedOpWhiteList[] = { // NOLINT
...
-  for (auto* op : view_.AllOps()) {
-    VLOG(4) << "Handle op " << cnt++ << ": " << op->Name();
-    if (FLAGS_enable_inplace_whitelist && !whitelist_.count(op->Name()))
-      continue;
-    TryInplaceOpInputOutput(op, graph);
-  }
-}
+  for (auto *node : ops) {
+    if (!node->IsOp()) continue;
+    // avoid optimize the variable used in sub-blocks
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/inplace_op_pass.cc]: Grammar: "avoid doing"

File: paddle/fluid/framework/details/inplace_op_pass.cc [hunk 3]
```diff
@@ -83,490 +78,374 @@ const std::string kInplacedOpWhiteList[] = { // NOLINT
...
-    graph->RemoveNode(node);
-    auto& nodes = var_nodes_.at(var);
-    // release unused var in graph. Because python side memory optimize
-    // may reused the var in same name, so we only clear the var node
-    // after current inplaced index.
-    nodes.erase(std::remove(nodes.begin(), nodes.end(), node), nodes.end());
+  if (all_out_nodes.empty()) {
+    ssa_map_.erase(out_var_name);
   }
+  op->Op()->RenameOutput(out_var_name, in_var_name);
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/inplace_op_pass.cc:289]: I saw you did Op()->Flush() after renaming at this function above. Do we need to do Flush() after renaming here?

File: paddle/fluid/framework/inplace_op_inference_test.cc [hunk 4]
```diff
@@ -202,7 +207,7 @@ ir::Node* GetNodeFromGraph(ir::Graph* g, std::string name) {
 
 std::unique_ptr<ir::Graph> test_SingleOpInplaceInToOut(
     std::unique_ptr<ir::Graph> g) {
-  std::unique_ptr<details::InplacePass> pass(new details::InplacePass());
+  auto pass = ir::PassRegistry::Instance().Get("inplace_pass");
```
Comment (zhhsplendid) [inline paddle/fluid/framework/inplace_op_inference_test.cc]: "auto pass = CreateInplacePass()" to de-duplicate code?

File: paddle/fluid/framework/details/reference_count_pass.cc [hunk 5]
```diff
@@ -267,6 +191,99 @@ static bool ShrinkNoNeedBufferVarOpDependency(
...
+/**
+ * Find the nearest downstream computation op handle. If the op is a
+ * computation op, just return itself.
+ */
+static ComputationOpHandle *FindNextComputationOpHandleOrReturnItself(
+    OpHandleBase *op, size_t scope_idx) {
+  std::queue<OpHandleBase *> q;
+  std::unordered_set<OpHandleBase *> visited;
+  q.push(op);
+  do {
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/reference_count_pass.cc]: Prefer "while" loop than "do while" loop. 

There is q.push(op) before the while loop. It's no harm to write
while (!q.empty()) {
    ...
}

File: paddle/fluid/framework/details/op_graph_view.h [hunk 6]
```diff
@@ -65,6 +65,7 @@ bool OpGraphView::VisitAllPendingOps(OpHandleBase *op,
         if (!callback(pending_op)) {
           return false;
         }
```
Comment (zhhsplendid) [inline paddle/fluid/framework/details/op_graph_view.h:67]: Prefer "while" loop than "do while" loop. 

There is q.push(op) before the while loop. It's no harm to write
while (!q.empty()) {
    ...
}

File: paddle/fluid/framework/details/inplace_op_pass.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/inplace_op_pass.cc]: Done.

File: paddle/fluid/framework/details/inplace_op_pass.cc (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/inplace_op_pass.cc]: Done.

File: paddle/fluid/framework/details/inplace_op_pass.cc (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/inplace_op_pass.cc:289]: Done.

File: paddle/fluid/framework/inplace_op_inference_test.cc (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/framework/inplace_op_inference_test.cc]: Done.

File: paddle/fluid/framework/details/reference_count_pass.cc (hunk 5 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/reference_count_pass.cc]: Done.

File: paddle/fluid/framework/details/op_graph_view.h (hunk 6 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/op_graph_view.h:67]: Done.

File: paddle/fluid/framework/details/inplace_op_pass.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/details/inplace_op_pass.cc]: Done.

File: paddle/fluid/framework/details/inplace_op_pass.cc (hunk 2 above)
Comment (zhhsplendid) [reply paddle/fluid/framework/details/inplace_op_pass.cc]: Oh >.< , I meant "avoid optimizing the variable ..."

**Review by zhhsplendid — DISMISSED** [review]
LGTM but please change the "avoid doing optimize" to "avoid optimizing" which I commented.

**Comment by sneaxiy** [comment]
@zhhsplendid Thanks. I have change "avoid dong optimize" to be "avoid optimizing".

**Review by chengduoZH — APPROVED** [review]
LGTM

**Pull request merged**